Dev

Romx98 opened the pull request:
First PR, completne get, xml parse, printAllVC

Romx98 committed:
first commit, created project

Romx98 committed:
first commit, created class

Romx98 committed:
update gitignore

Romx98 committed:
update gitignore

Romx98 committed:
add getResource, removeAllVC, printAllVC,...

Romx98 committed:
Osetrenie: Null attribute, http response error

Romx98 committed:
add test method docToStr,

kazooo commented on build.gradle:
```diff
@@ -0,0 +1,22 @@
+
+test {
+    useJUnitPlatform()
+}
```
Přidej new line - https://gist.github.com/camh-/1bebfcff1b0f814e9b191edc60d5206b

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/RestClient.java:
```diff
@@ -0,0 +1,117 @@
+    private final Transformer xmlTransformer;
+    private final XPathExpression xmlPathExp;
+
+    public RestClient(String fh, String fu, String fp)
```
Zkus jiný název, např. FedoraRestClient. Předpokládej, že budeš tohoto klienta znovu použivat v jiných skriptech, kde budou klienti pro Kramerius API, apod. Takže je potřeba dobrý název, podle kterého dokažeš těchto klientů rozlišit.

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/RestClient.java:
```diff
@@ -0,0 +1,117 @@
+        factory.setNamespaceAware(true);
+        xmlParser = factory.newDocumentBuilder();
+
+        XPathFactory xFactory = XPathFactory.newInstance();
+        XPath xmlPath = xFactory.newXPath();
+        xmlPathExp = xmlPath.compile("//*[local-name() = 'isMemberOfCollection']");
```
Pro parsování XML elementů bych udělal další pomocný servis. Těch XPathExpressions bude víc. Ale zatim at to zůstane takto.

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/RestClient.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    public Document removeAllVC(Document doc, String vc) throws XPathExpressionException {
+        if (doc == null) { return null; }
```
Snaž se vyhnout jednořádkovým if výrazům. U složitějších výrazu pak může byt složitý debugging apod

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/RestClient.java:
```diff
@@ -0,0 +1,117 @@
+    public Document removeAllVC(Document doc, String vc) throws XPathExpressionException {
+        if (doc == null) { return null; }
+
+        String uuidVC = getFullFormatVC(vc);
```
Nekontroluješ je-li vc null

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/RestClient.java:
```diff
@@ -0,0 +1,117 @@
+        NodeList nodes = (NodeList) xmlPathExp.evaluate(doc, XPathConstants.NODESET);
+        for (int i = 0; i < nodes.getLength(); i++) {
+            Node node = nodes.item(i);
+            if (node.getAttributes().item(0).getTextContent().equals(uuidVC) ||
```
```suggestion
            if (node.getAttributes().item(0).getTextContent().equals(uuidVC) ||
```
node.getAttributes() může vratit prazdný list, a .item(0) vyhodí ArrayOutOfBoxException
Mimo jiné, musiš vybrat attribut "resources", který nemusí byt první.

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/RestClient.java:
```diff
@@ -0,0 +1,117 @@
+        for (int i = 0; i < nodes.getLength(); i++) {
+            Node node = nodes.item(i);
+            if (node.getAttributes().item(0).getTextContent().equals(uuidVC) ||
+                node.getAttributes().item(0).getTextContent().isEmpty()) {
```
Pokud uuidVc je prazdný String, tak bude true ještě v první podmince, takže tento výraz je zbytečný.

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/RestClient.java:
```diff
@@ -0,0 +1,117 @@
+        xmlTransformer = tFactory.newTransformer();
+    }
+
+    public Document removeAllVC(Document doc, String vc) throws XPathExpressionException {
```
Název removeVc je trochu lepší podle mě, ale jak chceš, at to bude tak

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/RestClient.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    public void printAllVC(Document doc) throws XPathExpressionException {
+        if (doc == null) { return; }
```
Viz komentář výše

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/RestClient.java:
```diff
@@ -0,0 +1,117 @@
+        } catch (RuntimeException e) {
+            e.getMessage();
+        }
+        return null;
```
Nevracej null z funkcí, bad habbit. Koukni do Optional nebo vyhod výjimku.

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/RunRestClient.java:
```diff
@@ -0,0 +1,28 @@
+import javax.xml.transform.TransformerException;
+import javax.xml.xpath.XPathExpressionException;
+
+public class RunRestClient {
```
Název trídy nic neříká, změn na nějaký RemoveVcFromFOXML, aby název mohl říct co skriptu dělá.

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/RunRestClient.java:
```diff
@@ -0,0 +1,28 @@
+        RestClient restClient = new RestClient(fh, fu, fp);
+
+        Document doc = restClient.getFoxmlByUuid(uuid);
+        System.out.println(restClient.docToStr(doc));
```
V docToStr nekontroluješ je-li doc null -> NullPointerException

kazooo commented:
Názvy PR a commitů mohou byt informativnější, stejně tak i popisy. Snaž se názvy a popisy psat v angličtině aby nebyl mix jázyků.

Romx98 committed:
Closed request changed -> added new line in gradle.builder, renaming class names and method names (removeVc(..))

Romx98 committed:
Closed request changed -> edit XPath expression (../@resource), completing the removeVc method (checking if formal parameters are null; works with /@resource attributes; contains 2 conditions, the match uuidVc and the parents node is not null)

Romx98 committed:
Closed request changed -> added util Optional to check the null value, Methods that are prone to NullPointerExceptions have a return type Optional<...> (example. Optional<Document>)

Romx98 reviewed (commented):
Spravil som nejaké zmeny (pridal som Optional), čiže predpokladám, že by tam už nemal byť problém s return null; @kazooo

Romx98 committed:
Added new packages model, which contain enum class DataStreams (just contains two identifiers DC, RELS-EXT),

Romx98 committed:
probably unfinished (the first test did not work)... post method for identifiers DC, RELS-EX, added method for casting from doc to entity, create mimeType headers,... ;

Romx98 committed:
Unfinished... small changes in code, remove setDC method (because it's not used..), null return control, resolved issue with converting java.lang.Optional, but cannot connect to server (http status 500)...

kazooo commented:
Prosím zapisuj commit message jako

```
<stručný shrnující popis změn>

<podrobnější popis změn>
```
aby se ti v GUI nezobrazoval dlouhý popis, který je nutne otevřit abys zjistil o co jde.
Nezapisuj stav skriptu, hotový nebo ne. Když to přidáme do masteru musí byt už hotový.

Například

```
Extend FedoraClient API

 - add new REST API endpoint blablabla
 - refactor method()
```
Pak budeš mit krasně formatované commity a bude to cool  :)

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/FedoraRestClient.java:
```diff
@@ -0,0 +1,192 @@
+        httpEntity = new HttpEntity<>(httpHeaders);
+
+        DocumentBuilderFactory factory = DocumentBuilderFactory.newDefaultInstance();
+        //factory.setNamespaceAware(true);
```
Proč jsi to okomentoval?

Romx98 replied:
Mal som to odkomentovane, ked som pouzival XPath expression //*[local-name() = 'isMemberOfCollection'], lenze ked som skusil pouzit   //*[local-name() = 'isMemberOfCollection']/@resource tak mi to neslo... trosku som experimentoval a ked som to zakomentoval, tak mi to zrazu fungovalo

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/FedoraRestClient.java:
```diff
@@ -0,0 +1,192 @@
+
+    public Optional<Document> removeVc(Document doc, String vc)
+            throws XPathExpressionException {
+        if ((doc != null) && (vc != null)) {
```
Závorky kolem *doc* a *vc* jsou zbytečné, ale mohou byt

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/FedoraRestClient.java:
```diff
@@ -0,0 +1,192 @@
+            throws XPathExpressionException {
+        if ((doc != null) && (vc != null)) {
+            String uuidVc = getFullFormatVC(vc);
+            NodeList nodes = (NodeList) xmlPathExp.evaluate(doc, XPathConstants.NODESET);
```
Přejmenuj na attrubuteNodes, at tem název vic říká

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/FedoraRestClient.java:
```diff
@@ -0,0 +1,192 @@
+            NodeList nodes = (NodeList) xmlPathExp.evaluate(doc, XPathConstants.NODESET);
+
+            for (int i = 0; i < nodes.getLength(); i++) {
+                Node node = nodes.item(i);
```
node může byt attrNode

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/FedoraRestClient.java:
```diff
@@ -0,0 +1,192 @@
+                    Node nodeOwner = attr.getOwnerElement();
+                    if (nodeOwner.getParentNode() != null) {
+                        nodeOwner.getParentNode().removeChild(nodeOwner);
+                    }
```
Toto celé můžeš udělat jako samostatnou funkci

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/FedoraRestClient.java:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    public void setRelsExt(String uuid, Document relsExt)
+            throws IOException, TransformerException {
```
Na jeden řádek

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/FedoraRestClient.java:
```diff
@@ -0,0 +1,192 @@
+
+    public Optional<String> docToStr(Document doc) throws TransformerException {
+        String str = null;
+
```
Zbytečně

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/FedoraRestClient.java:
```diff
@@ -0,0 +1,192 @@
+
+    private Optional<HttpEntity<Object>> docStrEntity(Document doc, String mimeType) throws TransformerException {
+        HttpEntity<Object> entity = null;
+
```
Zbytečně

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/FedoraRestClient.java:
```diff
@@ -0,0 +1,192 @@
+
+    private Optional<Document> getFedoraResource(String url) {
+        Document responseDoc = null;
+
```
Zbytečně

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/FedoraRestClient.java:
```diff
@@ -0,0 +1,192 @@
+        return headers;
+    }
+
+    public Optional<String> docToStr(Document doc) throws TransformerException {
```
Když je to private funkce (asi by měla byt private), tak bych tady čekal, že předtím něž pošleš doc na konverzí, tak ho zkontroluješ, takže Optional není potřeba

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/FedoraRestClient.java:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    private void setDataStream(String uuid, DataStreams ds, HttpEntity<Object> entity)
+            throws IOException {
```
Na jeden řádek

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/FedoraRestClient.java:
```diff
@@ -0,0 +1,192 @@
+            throws IOException {
+        if (entity != null) {
+            String url = fedoraHost + "/objects/" + uuid + "/datastreams/" + ds.name;
+
```
Zbytečně

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/FedoraRestClient.java:
```diff
@@ -0,0 +1,192 @@
+            url = buildUri(url, uriParam);
+
+            ResponseEntity<String> responseEntity = restTemplate.postForEntity(url, entity, String.class);
+
```
Zbytečně

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/RemoveVcFromFOXML.java:
```diff
@@ -0,0 +1,41 @@
+        FedoraRestClient fedoraRestClient = new FedoraRestClient(fh, fu, fp);
+
+        Optional<Document> doc = fedoraRestClient.getFoxmlByUuid(uuid);
+        //System.out.println(fedoraRestClient.docToStr(doc));
```
Okomentovaný kodu ideálně nemá byt v commitech

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/RemoveVcFromFOXML.java:
```diff
@@ -0,0 +1,41 @@
+        }
+
+        if (doc.isPresent()) {
+            fedoraRestClient.setRelsExt(uuid, doc.get());
```
Ziskáváš full FOXML, a posíláš ho jako RELS-EXT, Fedora ti to neakceptuje. Zpracuj jen RELS-EXT, tady to bohatě stačí.

kazooo commented on src/main/java/cz/mzk/fedora/documentcorrection/RemoveVcFromFOXML.java:
```diff
@@ -0,0 +1,41 @@
+
+
+
+
```
Zbytečně

kazooo commented on src/main/java/cz/mzk/fedora/model/DataStreams.java:
```diff
@@ -0,0 +1,38 @@
+
+    DC("DC", "text/xml", "false", "X", "A"),
+    RELS_EXT("RELS-EXT", "application/rdf+xml" , "true", "X", "A");
+//
```
?

kazooo commented on src/main/java/cz/mzk/fedora/model/DataStreams.java:
```diff
@@ -0,0 +1,38 @@
+    public String controlGroup;
+    public String state;
+
+
```
Možná ne vždy to jde, ale snaž se mezi logickými bloky koda dodržovat pouze jeden řádek.

kazooo commented on src/main/java/cz/mzk/fedora/model/DataStreams.java:
```diff
@@ -0,0 +1,38 @@
+
+    public Optional<DataStreams> getDataStreams(String dsName) {
+        DataStreams dataStreams = null;
+
```
Zbytečně

kazooo commented on src/main/java/cz/mzk/fedora/model/DataStreams.java:
```diff
@@ -0,0 +1,38 @@
+        }
+        return Optional.ofNullable(dataStreams);
+    }
+
```
Zbytečně

kazooo reviewed (changes requested):
Oprav to co tady už máš a už neposilej nic nového, máme už velmi dlouhý log

Romx98 committed:
Summary description of changes:
 - added a method to obtain the RELS-EXT identifier
 - the first test ran
 - the code needs to be modified

Romx98 committed:
Summary descriptions of changes:
 - completed script
 - delete blank lines
 - adding the removeChildren helper method to the removeVc method
 - modifying methods using the Optional utility (added Optional.empty())
